### ISSUE TITLE
Keras load_model fails after training

### ISSUE THREAD

**Issue by AahanSingh** [fchollet/keras#5442]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

I tried the code given here:  https://github.com/fchollet/keras/issues/2295

I am unable to load my model after training it. I am getting the following error:

`ValueError: Optimizer weight shape (3, 3, 512, 512) not compatible with provided weight shape (256, 43)`

I tried the solution here: https://github.com/fchollet/keras/issues/3964 using HDFView and deleting the optimiser weights and then reloading. But then I get the error:

```
ValueError: ('shapes (10,4224) and (1128,256) not aligned: 4224 (dim 1) != 1128 (dim 0)', (10, 4224), (1128, 256))
Apply node that caused the error: Dot22(Reshape{2}.0, lstm_2_W_i)
Toposort index: 249
Inputs types: [TensorType(float32, matrix), TensorType(float32, matrix)]
Inputs shapes: [(10, 4224), (1128, 256)]
Inputs strides: [(16896, 4), (1024, 4)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[Elemwise{Add}[(0, 0)](Dot22.0, InplaceDimShuffle{x,0}.0)]]
```
I also tried the solution by barrykui here: https://github.com/fchollet/keras/issues/4044 saving the architecture to JSON and saving weights separately and then loading but even that fails.

**Comment by AahanSingh** [comment]
I was able to devise a workaround. I followed the steps below:

1. Create the Sequential model.
Example: 
```
model = Sequential()
model.add(...)
model.add(...)

model.compile(...)
model.fit(...)
```

2. After training save only model weights using model.save_weights()
Example:
`model.save_weights(SaveLocation)`
3. To load the model weights, create the model programmatically just as in step 1 but do not use the model.compile function.
Example:
```
model = Sequential()
model.add(...)
model.add(...)
model.load_weights(weightFile)
```

4. The weights are now loaded successfully.

**Comment by AahanSingh** [comment]
I reopened this issue just in case someone else is able to provide a better solution.

**Comment by WarpTime** [comment]
Another solution is to name each layer in the model.
This helps if you're using checkpointing or something similar, where you cannot save the weights separately.
For ex: model.add(LSTM(10, name='lstm1'))
This worked for me in keras 1.1.1 and theano 0.8.2

**Comment by svanimisetti** [comment]
I observed this issue on Keras 1.2.2 with TensorFlow 0.12.1 backend on Win64 (. The best solution so far has been from @AahanSingh (without needing to upgrade my setup). Other solutions have alluded to naming each layer and retraining before saving model JSON and weights, which for me was not a convenient option. I have redefined the model as per the suggestion - and this seems to be the simplest workaround till date (at least without having to upgrade my setup).
